apps/lvgl_scanner/src/main.c

```diff
@@ -0,0 +1,333 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+#include "sysinit/sysinit.h"
+#include "os/os.h"
+#include "console/console.h"
+#include "host/ble_hs.h"
+#include "log/log.h"
+#include "lvgl.h"
+
+#define MAX_DEVICES         30
+#define DEVICE_ID_NONE      -1
+
+lv_style_t style;
+lv_obj_t *entries_list;
+
+struct device_data {
+    int8_t last_rssi;
+    ble_addr_t addr;
+    lv_obj_t *list_entry;
+};
+
+static struct device_data devices[MAX_DEVICES];
+
+int
+rssi_compare(const void *p1, const void *p2)
+{
+    struct device_data *d1 = (struct device_data *) p1;
+    struct device_data *d2 = (struct device_data *) p2;
+
+    return (d2->last_rssi - d1->last_rssi);
+}
+
+static void
+sort_devices(void)
+{
+    uint8_t i;
+    uint8_t n = lv_obj_get_child_cnt(entries_list);
+
+    qsort(devices, n, sizeof(*devices), rssi_compare);
+
+    for (i = 0; i < n; i++) {
+        lv_obj_move_to_index(devices[i].list_entry, i);
+    }
+}
+
+static int
+get_device_id_by_addr(uint8_t *addr)
+{
+    uint8_t id;
+    uint8_t n = lv_obj_get_child_cnt(entries_list);
+    assert (n <= MAX_DEVICES);
+
+    for (id = 0; id < n; id++) {
+        if (!memcmp(addr, devices[id].addr.val, 6)) {
+            return id;
+        }
+    }
+    return DEVICE_ID_NONE;
+}
+
+/* This fills entry labels which are going to be displayed. Takes is_new flag as argument.
+ * If flag is set, then new labels will be created, otherwise existing labels will
+ * be updated */
+static void
+fill_device_labels(struct ble_gap_ext_disc_desc *desc, uint8_t id, uint8_t is_new)
```

Comment:
probably it makes more sense to have separate function to initialize empty object for new device (i.e. `add_new_device`) and then call update all labels which can be used for both existing device and the one that we just created (i.e. `update_device`)